apiserver/facades/client/machinemanager/machinemanager.go

```diff
@@ -372,7 +373,29 @@ func (mm *MachineManagerAPI) destroyMachine(args params.Entities, force, keep bo
 				logger.Warningf("could not keep instance for machine %v: %v", machineTag.Id(), err)
 			}
 		}
-		var info params.DestroyMachineInfo
+		info := params.DestroyMachineInfo{
+			MachineId: machineTag.Id(),
+		}
+
+		containers, err := machine.Containers()
+		if err != nil {
+			return fail(err)
+		}
+		containerEntities := params.Entities{Entities: make([]params.Entity, len(containers))}
+		for i, container := range containers {
+			containerEntities.Entities[i] = params.Entity{Tag: names.NewMachineTag(container).String()}
+		}
+		if len(containers) > 0 {
+			if !force {
+				return fail(errors.Errorf("machine %v has container machines %q", machineTag.Id(), strings.Join(containers, ", ")))
+			}
+			containerResults, err := mm.destroyMachine(containerEntities, force, keep, maxWait)
```

Comment:
A container cannot have a container, it's a bit of wasted effort to call mm.destroyMachine here.

Reply:
The fact this can technically handle infinite recursion is a nice addition. However, this is required to destroy containers hosted on machines (or rather, to explicitly destroy them, which would be done beforehand as a side effect to `machine.ForceDestory()`, but in that case we would lose the results (We wouldn't know the container `MachineId`s, the units destroyed, etc.)

Perhaps an argument could be made this data shouldn't be collected with recursion like this, but I feel that would add unnecessary complexity